File: vesopa_server/src/fonts.js
/**
 * Fonts — the lettering a venue's tills wear.
 *
 * Two kinds, one list. Sixteen built-in families are files in the tree, fetched
 * once by tool/fetch_fonts.js and served from public/assets/fonts; a venue's
 * own uploads are rows in epos_fonts with files under public/uploads/fonts.
 * Everything downstream — the screen editor's picker, the till's cache, a
 * button's `fontFamily` — sees one flat list of families keyed by slug and does
 * not care which kind a font is, except that a built-in cannot be deleted.
 *
 * WHY THE FILES LIVE HERE AND NOT ON A CDN.
 *
 * The obvious build is a <link> to fonts.googleapis.com in the back office and
 * a Google Fonts URL on the till. It is wrong for the till twice over. A till
 * is offline-first by design — it takes money through a broadband outage, which
 * is most of the reason it exists — and a button whose lettering arrives over
 * the internet is a button that changes shape when the line drops. And Flutter's
 * FontLoader reads ttf and otf; Google serves woff2 to anything modern, so the
 * till would have to lie about its user agent to get a file it could use.
 *
 * So the till downloads fonts from this back office, over the same connection
 * it already fetches products and screens on, and caches them on disk. First
 * run needs the network. Nothing after it does.
 *
 * WHY .woff2 IS REFUSED ON UPLOAD.
 *
 * For the same reason, and it is worth refusing loudly rather than storing a
 * file that works in the editor's preview and silently does nothing on the
 * terminal. A manager who uploads their brand woff2, sees it in the back
 * office, and finds plain lettering at the counter has been told a lie by this
 * software. The error names the formats that work.
 *
 * ROUTING, and the trap it avoids.
 *
 *   fontsRoutes      the back office, on a session token.
 *   tillFontRoutes   the tills. The read is unauthenticated and scoped by an
 *                    `office` query, exactly as /till/screens is. The *write*
 *                    is not: uploading a font puts a file on our disk, so it
 *                    takes a terminal token, which is the strongest thing a
 *                    till has to present.
 *
 * As with the kitchen's and the screens' routers, these two must never share a
 * path — the back office's `auth` in front of a till read would take the
 * lettering off every terminal in every venue at once. They do not: the till's
 * live under /till/.
 */

'use strict';

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

const express = require('express');
const multer = require('multer');

const { requireAuth, requireTerminal } = require('./auth');

const BUILT_IN_DIR = path.join(__dirname, '..', 'public', 'assets', 'fonts');
const UPLOAD_DIR = path.join(__dirname, '..', 'public', 'uploads', 'fonts');

/** Weights a family may have. Regular and bold; see tool/fetch_fonts.js. */
const WEIGHTS = [400, 700];

/**
 * A font file is at most this. Four megabytes is a generous CJK face and about
 * twenty times a Latin one — and it is also what a till has to pull down over a
 * venue's broadband before it can draw a button, which is the real ceiling.
 */
const MAX_BYTES = 4 * 1024 * 1024;

/**
 * The built-in families, read once at startup from the manifest the fetch tool
 * wrote. Read once rather than per request because these files cannot change
 * without a deploy, and a deploy restarts this process.
 *
 * A missing or unreadable manifest is not fatal. It means no built-in fonts —
 * the venue keeps its uploads, every till falls back to the app's own typeface,
 * and nothing 500s. A back office that will not start because a font list is
 * malformed is a worse outcome than one that letters things plainly.
 */
const BUILT_INS = (() => {
  try {
    const raw = fs.readFileSync(path.join(BUILT_IN_DIR, 'catalogue.json'), 'utf8');
    const parsed = JSON.parse(raw);
    if (!Array.isArray(parsed)) return [];
    return parsed
      .filter((f) => f && f.slug && f.family && Array.isArray(f.faces))
      .map((f) => ({
        slug: String(f.slug),
        family: String(f.family),
        builtIn: true,
        faces: f.faces
          .filter((face) => WEIGHTS.includes(Number(face.weight)))
          .map((face) => ({
            weight: Number(face.weight),
            url: `/assets/fonts/${f.slug}/${face.file}`,
            bytes: Number(face.bytes) || 0,
          }))
          .sort((a, b) => a.weight - b.weight),
      }))
      .filter((f) => f.faces.length);
  } catch {
    return [];
  }
})();

const BUILT_IN_SLUGS = new Set(BUILT_INS.map((f) => f.slug));

/**
 * A family name reduced to the key a button stores.
 *
 * Buttons and till settings hold the slug, never the display name, so a venue
 * can rename "Brand Sans" to "Aishi Display" without every key it letters
 * falling back to plain. That is also why the slug is derived once, on the
 * first upload of a family, and never recomputed from a changed name.
 */
function slugify(name) {
  return String(name || '')
    .toLowerCase()
    .normalize('NFKD')
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 64);
}

/**
 * Is this actually a font, by its first four bytes?
 *
 * The extension is what the browser sends and is worth nothing: a .ttf is
 * whatever somebody renamed. These four signatures are the whole of what
 * FontLoader can read.
 *
 *   0x00010000  TrueType outlines — the common .ttf
 *   'true'      TrueType, the older Apple tag
 *   'ttcf'      a TrueType collection
 *   'OTTO'      OpenType with CFF outlines — .otf
 *
 * Deliberately NOT here: 'wOFF' and 'wOF2'. They are real fonts and the browser
 * would render them; the till could not. See the header.
 */
function fontKind(buffer) {
  if (!buffer || buffer.length < 4) return null;
  const tag = buffer.subarray(0, 4).toString('latin1');
  if (tag === 'OTTO') return 'otf';
  if (tag === 'true' || tag === 'ttcf') return 'ttf';
  if (buffer.readUInt32BE(0) === 0x00010000) return 'ttf';
  if (tag === 'wOFF' || tag === 'wOF2') return 'woff';
  return null;
}

/**
 * Uploads land in memory, not on disk.
 *
 * multer's disk storage would write the file before anything had looked at it,
 * so a rejected upload leaves rubbish in public/uploads for somebody to find
 * later — and public/uploads is served statically, which makes "we stored a
 * file we refused" worse than untidy. Four megabytes through RAM is nothing.
 */
const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: MAX_BYTES, files: 1 },
});

/** The rows a venue uploaded, as families. */
async function uploadedFor(pool, office) {
  const [rows] = await pool.query(
    `SELECT slug, family, weight, file_name, byte_size
       FROM epos_fonts
      WHERE office = ?
      ORDER BY family, weight`,
    [office]
  );

  const bySlug = new Map();
  for (const row of rows) {
    if (!bySlug.has(row.slug)) {
      bySlug.set(row.slug, {
        slug: row.slug,
        family: row.family,
        builtIn: false,
        faces: [],
      });
    }
    bySlug.get(row.slug).faces.push({
      weight: Number(row.weight),
      url: `/uploads/fonts/${row.file_name}`,
      bytes: Number(row.byte_size) || 0,
    });
  }
  return [...bySlug.values()];
}

/**
 * Every font this venue can letter a till in.
 *
 * Built-ins first and in the order the fetch tool listed them — which is
 * roughly "what a till looks best in" rather than alphabetical, because the
 * first few entries are the ones a manager in a hurry picks from. The venue's
 * own follow, sorted by name, because by then they are looking for one.
 */
async function catalogueFor(pool, office) {
  const own = await uploadedFor(pool, office);
  own.sort((a, b) => a.family.localeCompare(b.family));
  // A venue that uploads a family named like a built-in gets its own: it is on
  // their tills because they put it there, and shadowing is a less surprising
  // outcome than an upload that appears to do nothing.
  const shadowed = new Set(own.map((f) => f.slug));
  return [...BUILT_INS.filter((f) => !shadowed.has(f.slug)), ...own];
}

/**
 * Store one uploaded face, replacing whatever that weight was before.
 *
 * Shared by the back office's route and the till's, because "a till uploaded
 * it" and "a manager uploaded it" differ only in who is named in uploaded_by.
 * Two copies of this would be two chances to forget the signature check.
 */
async function storeFace(pool, { office, file, family, weight, by }) {
  const kind = fontKind(file.buffer);
  if (kind === 'woff') {
    const err = new Error(
      'A .woff or .woff2 font works in a browser but not on a till. ' +
        'Upload the .ttf or .otf the foundry supplies alongside it.'
    );
    err.status = 400;
    throw err;
  }
  if (!kind) {
    const err = new Error('That file is not a TrueType or OpenType font.');
    err.status = 400;
    throw err;
  }

  const name = String(family || '').trim().slice(0, 64);
  const slug = slugify(name);
  if (!name || !slug) {
    const err = new Error('Give the font a name.');
    err.status = 400;
    throw err;
  }
  if (BUILT_IN_SLUGS.has(slug)) {
    const err = new Error(
      `“${name}” is one of the built-in fonts. Give yours a different name so ` +
        'the two can be told apart on the till.'
    );
    err.status = 409;
    throw err;
  }

  const w = WEIGHTS.includes(Number(weight)) ? Number(weight) : 400;

  // Named for what it is rather than for what the browser called it, so a
  // directory listing reads `brand-sans-700-....ttf` and not `blob`. The random
  // tail is not decoration: every office writes into one directory, two venues
  // may both have a "Brand Sans", and `brand-sans-700.ttf` would mean whichever
  // uploaded second silently replaced the other venue's font. It also makes the
  // URL of a replaced face different from the one it replaced, which is what
  // stops a browser and a till serving yesterday's file out of cache.
  const tail = crypto.randomBytes(6).toString('hex');
  const fileName = `${slug}-${w}-${tail}.${kind}`;

  fs.mkdirSync(UPLOAD_DIR, { recursive: true });
  fs.writeFileSync(path.join(UPLOAD_DIR, fileName), file.buffer);

  // The old file, if this weight is being replaced. Read before the write so a
  // failed write leaves the venue with the font it had.
  const [[previous]] = await pool.query(
    'SELECT file_name FROM epos_fonts WHERE office = ? AND slug = ? AND weight = ?',
    [office, slug, w]
  );

  await pool.query(
    `INSERT INTO epos_fonts (office, family, slug, weight, file_name, byte_size, uploaded_by)
     VALUES (?, ?, ?, ?, ?, ?, ?)
     ON DUPLICATE KEY UPDATE
       family = VALUES(family),
       file_name = VALUES(file_name),
       byte_size = VALUES(byte_size),
       uploaded_by = VALUES(uploaded_by)`,
    [office, name, slug, w, fileName, file.buffer.length, by || null]
  );

  if (previous && previous.file_name && previous.file_name !== fileName) {
    // Best effort. An orphaned font file is a few hundred kilobytes; a failed
    // upload because the old file was already gone is a manager stuck.
    try {
      fs.unlinkSync(path.join(UPLOAD_DIR, previous.file_name));
    } catch {
      /* already gone */
    }
  }

  return { slug, family: name, weight: w, bytes: file.buffer.length };
}

/** Remove a venue's font, file and all. Built-ins refuse. */
async function removeFamily(pool, office, slug) {
  if (BUILT_IN_SLUGS.has(slug)) {
    const err = new Error('The built-in fonts cannot be removed.');
    err.status = 400;
    throw err;
  }

  const [rows] = await pool.query(
    'SELECT file_name FROM epos_fonts WHERE office = ? AND slug = ?',
    [office, slug]
  );
  if (!rows.length) return false;

  await pool.query('DELETE FROM epos_fonts WHERE office = ? AND slug = ?', [
    office,
    slug,
  ]);
  for (const row of rows) {
    try {
      fs.unlinkSync(path.join(UPLOAD_DIR, row.file_name));
    } catch {
      /* already gone */
    }
  }
  return true;
}

// ---------------------------------------------------------------------------
// The back office
// ---------------------------------------------------------------------------

function fontsRoutes({ pool, secret, broadcast }) {
  const router = express.Router();
  const auth = requireAuth(secret);

  async function tenantEmail(req) {
    if (req.user.officeId) {
      const [[office]] = await pool.query(
        'SELECT contact_email FROM offices WHERE id = ?',
        [req.user.officeId]
      );
      if (office) return office.contact_email;
    }
    return req.user.email;
  }

  /**
   * Tell this venue's tills the font list moved.
   *
   * Reusing the `screens` push rather than minting a `fonts` one: a till that
   * hears "screens" already re-reads its layout, and the fonts a layout needs
   * are fetched in the same breath. A second message would mean a second
   * handler on the till and a version of the till that ignores it.
   */
  function pushed(office) {
    broadcast({ type: 'screens', office }, { office });
  }

  router.get('/fonts', auth, async (req, res, next) => {
    try {
      const office = await tenantEmail(req);
      res.json({ fonts: await catalogueFor(pool, office) });
    } catch (e) {
      next(e);
    }
  });

  /**
   * Upload one face of one family.
   *
   * `family` and `weight` are fields beside the file rather than parsed out of
   * the filename. A foundry's file is called `BrandSans-Bd_v2_FINAL.ttf` and
   * guessing "Bd" means 700 is the kind of cleverness that letters half a
   * venue's buttons in the wrong weight.
   */
  router.post('/fonts', auth, (req, res, next) => {
    upload.single('font')(req, res, async (err) => {
      if (err) {
        return res.status(400).json({
          error:
            err.code === 'LIMIT_FILE_SIZE'
              ? `That font is larger than ${MAX_BYTES / 1024 / 1024} MB.`
              : 'That upload could not be read.',
        });
      }
      if (!req.file) return res.status(400).json({ error: 'No font was sent.' });

      try {
        const office = await tenantEmail(req);
        const stored = await storeFace(pool, {
          office,
          file: req.file,
          family: req.body.family,
          weight: req.body.weight,
          by: req.user.email,
        });
        pushed(office);
        res.status(201).json(stored);
      } catch (e) {
        if (e.status) return res.status(e.status).json({ error: e.message });
        next(e);
      }
    });
  });

  router.delete('/fonts/:slug', auth, async (req, res, next) => {
    try {
      const office = await tenantEmail(req);
      const gone = await removeFamily(pool, office, String(req.params.slug));
      if (!gone) return res.status(404).json({ error: 'No such font.' });
      pushed(office);
      res.json({ ok: true });
    } catch (e) {
      if (e.status) return res.status(e.status).json({ error: e.message });
      next(e);
    }
  });

  return router;
}

// ---------------------------------------------------------------------------
// The tills
// ---------------------------------------------------------------------------

function tillFontRoutes({ pool, secret, broadcast }) {
  const router = express.Router();

  /**
   * What this venue's tills may letter themselves in.
   *
   * Unauthenticated and scoped by `office`, as /till/screens is. The till
   * fetches this on startup, downloads anything it has not cached, and keeps
   * working from the cache when this is unreachable.
   */
  router.get('/till/fonts', async (req, res, next) => {
    try {
      const office = String(req.query.office || '').trim();
      if (!office) return res.status(400).json({ error: 'office is required' });
      res.json({ fonts: await catalogueFor(pool, office) });
    } catch (e) {
      next(e);
    }
  });

  /**
   * A font uploaded from the counter.
   *
   * On a terminal token, not on `office` alone. Every other till route here is
   * a read, and an unauthenticated read scoped by a venue's email address is a
   * reasonable trade; an unauthenticated *write* that puts a file on our disk
   * is not, whatever it is scoped by.
   */
  router.post(
    '/till/fonts',
    requireTerminal(secret),
    (req, res, next) => {
      upload.single('font')(req, res, async (err) => {
        if (err) {
          return res.status(400).json({
            error:
              err.code === 'LIMIT_FILE_SIZE'
                ? `That font is larger than ${MAX_BYTES / 1024 / 1024} MB.`
                : 'That upload could not be read.',
          });
        }
        if (!req.file) return res.status(400).json({ error: 'No font was sent.' });

        try {
          const stored = await storeFace(pool, {
            office: req.office,
            file: req.file,
            family: req.body.family,
            weight: req.body.weight,
            by: req.terminal.name || 'a till',
          });
          // Every other till in the venue gets it too. A font installed at one
          // counter and missing at the next is the fault this push prevents.
          broadcast({ type: 'screens', office: req.office }, { office: req.office });
          res.status(201).json(stored);
        } catch (e) {
          if (e.status) return res.status(e.status).json({ error: e.message });
          next(e);
        }
      });
    }
  );

  /**
   * The font this venue's tills wear, changed from a till.
   *
   * A venue-wide setting written from a counter, which is unusual enough to say
   * why: the manager who wants to see whether the brand font is legible across
   * a bar is standing at the bar, not at a desk. Making them walk to an office,
   * change it, and walk back to look is how a font gets picked badly.
   *
   * Terminal token, and it writes exactly one column. Every till in the venue
   * is told, because a venue lettered two ways is worse than a venue lettered
   * in the wrong font.
   */
  router.put('/till/font', requireTerminal(secret), async (req, res, next) => {
    try {
      const slug = String(req.body?.fontFamily ?? '')
        .trim()
        .toLowerCase()
        .replace(/[^a-z0-9-]/g, '')
        .slice(0, 64);

      await pool.execute(
        `INSERT INTO epos_till_settings (office, font_family)
         VALUES (?, ?)
         ON DUPLICATE KEY UPDATE font_family = VALUES(font_family)`,
        [req.office, slug || null]
      );

      // 'till-settings', not 'screens': this is a settings row, and every till
      // already re-reads that row on this message. The font list has not
      // changed, only which one is chosen.
      broadcast({ type: 'till-settings' }, { office: req.office });
      res.json({ fontFamily: slug || null });
    } catch (e) {
      next(e);
    }
  });

  return router;
}

module.exports = {
  fontsRoutes,
  tillFontRoutes,
  // Exported for the tests, and for anything that needs to know what a font is
  // without going through a route.
  slugify,
  fontKind,
  catalogueFor,
  BUILT_INS,
  BUILT_IN_SLUGS,
  WEIGHTS,
  MAX_BYTES,
};

File: vesopa_epos/lib/data/fonts.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter/services.dart' show FontLoader;
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:http/http.dart' as http;
import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../config/constants.dart';
import '../main.dart';
import 'screens.dart';

/// The lettering a venue's tills wear.
///
/// Sixteen families ship with the back office and a venue may upload its own;
/// either way the files come from the venue's own back office, over the same
/// connection this till already fetches products and screens on. See
/// `vesopa_server/src/fonts.js` for why they are not fetched from Google.
///
/// **THE OFFLINE RULE, WHICH IS THE WHOLE DESIGN.**
///
/// A till takes money through a broadband outage. That is most of the reason it
/// exists. So a font is downloaded once, written to this terminal's own disk,
/// and registered from disk on every start after that — the network is needed
/// the first time a venue picks a font and never again. A font that has not
/// arrived yet is not an error and is not worth a dialog: the key is lettered
/// in the app's own typeface, which is exactly what every key looked like
/// before this feature existed, and it picks up the venue's font the next time
/// the till can reach the office.
///
/// Nothing here is allowed to throw into a sale. Every failure — no network, a
/// half-written file, a font file that is not a font, a disk that will not be
/// written to — comes out as plain lettering.

/// One weight of one family.
class FontFace {
  const FontFace({required this.weight, required this.url});

  final int weight;

  /// Absolute, resolved against the back office at parse time. The server
  /// sends `/assets/fonts/...` or `/uploads/fonts/...`, which is not something
  /// [http.get] can be handed.
  final String url;

  factory FontFace.fromJson(Map<String, dynamic> j) => FontFace(
    weight: (j['weight'] as num?)?.toInt() ?? 400,
    url: _absolute('${j['url'] ?? ''}'),
  );

  Map<String, dynamic> toJson() => {'weight': weight, 'url': url};

  static String _absolute(String path) {
    if (path.startsWith('http')) return path;
    return '${Api.resolvedBase}$path';
  }

  /// What this face is saved as on disk. The extension matters: the engine
  /// reads the bytes, but a `.ttf` holding OpenType outlines is the kind of
  /// thing that is impossible to diagnose from a directory listing.
  String fileName(String slug) =>
      '$slug-$weight${p.extension(Uri.parse(url).path).isEmpty ? '.ttf' : p.extension(Uri.parse(url).path)}';
}

/// One family — a built-in, or one this venue uploaded.
class VenueFont {
  const VenueFont({
    required this.slug,
    required this.family,
    required this.builtIn,
    required this.faces,
  });

  /// The stable key. A button stores this, never the display name, so renaming
  /// a font in the back office does not un-letter every key using it.
  final String slug;

  /// What a manager called it. Shown in Settings; never used to match.
  final String family;

  final bool builtIn;
  final List<FontFace> faces;

  /// The family name this is registered under with the engine.
  ///
  /// Prefixed so a venue that uploads a font called "Roboto" cannot collide
  /// with anything the app itself ships, and so a slug can never accidentally
  /// name a system font — `fontFamily: 'Arial'` would silently work on Windows
  /// and silently not on Android, which is the worst of both.
  String get engineFamily => 'vf-$slug';

  factory VenueFont.fromJson(Map<String, dynamic> j) => VenueFont(
    slug: '${j['slug'] ?? ''}',
    family: '${j['family'] ?? ''}',
    builtIn: j['builtIn'] == true,
    faces: [
      for (final f in (j['faces'] as List? ?? const []))
        FontFace.fromJson(f as Map<String, dynamic>),
    ],
  );

  Map<String, dynamic> toJson() => {
    'slug': slug,
    'family': family,
    'builtIn': builtIn,
    'faces': [for (final f in faces) f.toJson()],
  };
}

/// Every font this venue may letter a till in.
class FontLibrary {
  const FontLibrary(this.fonts, {this.installed = const <String>{}});

  static const FontLibrary empty = FontLibrary(<VenueFont>[]);

  final List<VenueFont> fonts;

  /// The slugs whose files are on this terminal's disk and registered with the
  /// engine. A font in [fonts] but not in here is one the venue has chosen and
  /// this till has not managed to download yet.
  final Set<String> installed;

  VenueFont? bySlug(String? slug) {
    if (slug == null || slug.isEmpty) return null;
    for (final f in fonts) {
      if (f.slug == slug) return f;
    }
    return null;
  }

  /// The family to hand a `TextStyle`, or null for the app's own typeface.
  ///
  /// The one place that decides, and it says no in three cases that all look
  /// the same to a clerk and are worth keeping apart in the code:
  ///
  ///  * nothing was asked for — most keys, most of the time;
  ///  * a font was asked for that this venue no longer has, because it was
  ///    deleted in the back office after this layout was cached;
  ///  * a font was asked for that is not on this terminal's disk yet.
  ///
  /// The last one is the reason this is a lookup rather than a string
  /// concatenation. Handing the engine a family it has never been given resolves
  /// to *something* — usually the platform default, sometimes a fallback with
  /// different metrics — and a key that silently changes shape is harder to
  /// explain than one that has not changed at all.
  String? familyFor(String? slug) {
    if (slug == null || slug.isEmpty) return null;
    if (!installed.contains(slug)) return null;
    return bySlug(slug)?.engineFamily;
  }

  FontLibrary withInstalled(Set<String> slugs) =>
      FontLibrary(fonts, installed: slugs);

  factory FontLibrary.fromJson(Map<String, dynamic> j) => FontLibrary([
    for (final f in (j['fonts'] as List? ?? const []))
      VenueFont.fromJson(f as Map<String, dynamic>),
  ]);

  Map<String, dynamic> toJson() => {
    'fonts': [for (final f in fonts) f.toJson()],
  };
}

class FontsRepository {
  FontsRepository({required this.apiBase, http.Client? client})
    : _client = client ?? http.Client();

  final String apiBase;
  final http.Client _client;

  static const _key = 'vesopa_till_fonts';

  /// Families already handed to the engine in this run of the app.
  ///
  /// Static, and deliberately: [FontLoader] registers with the engine, which
  /// outlives any provider. Registering the same family twice is not an error
  /// but it re-parses the file and appends a second copy of every face, and a
  /// till left running for a week through a hundred catalogue pushes would do
  /// that a hundred times.
  static final Set<String> _registered = <String>{};

  /// The list, from the office if it can be reached and from the cache if not.
  ///
  /// [wanted] is the set of slugs this venue is actually lettering something
  /// in. Only those are downloaded, and that matters more than it looks: the
  /// catalogue is sixteen built-in families before a venue has uploaded
  /// anything, and fetching the lot would be four and a half megabytes over a
  /// venue's broadband to draw one screen in one of them. Empty means download
  /// nothing, which is the right answer for a venue that has chosen no font.
  Future<FontLibrary> load(String office, {Set<String> wanted = const {}}) async {
    FontLibrary library;
    try {
      final res = await _client
          .get(Uri.parse('$apiBase/api/till/fonts?office=$office'))
          .timeout(const Duration(seconds: 8));
      if (res.statusCode == 200) {
        library = FontLibrary.fromJson(
          jsonDecode(res.body) as Map<String, dynamic>,
        );
        final prefs = await SharedPreferences.getInstance();
        await prefs.setString(_key, jsonEncode(library.toJson()));
      } else {
        library = await _cached();
      }
    } catch (_) {
      library = await _cached();
    }

    return library.withInstalled(await _install(library, wanted));
  }

  Future<FontLibrary> _cached() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final raw = prefs.getString(_key);
      if (raw == null || raw.isEmpty) return FontLibrary.empty;
      return FontLibrary.fromJson(jsonDecode(raw) as Map<String, dynamic>);
    } catch (_) {
      return FontLibrary.empty;
    }
  }

  /// Where the files live. Application support, not a cache directory: a cache
  /// is a place the operating system may empty, and a till that loses its
  /// lettering because Windows wanted the space — offline, mid-service — is a
  /// till nobody can fix at the counter.
  Future<Directory> _dir() async {
    final base = await getApplicationSupportDirectory();
    final dir = Directory(p.join(base.path, 'fonts'));
    if (!dir.existsSync()) dir.createSync(recursive: true);
    return dir;
  }

  /// Download anything missing, then register every family that is complete.
  ///
  /// Returns the slugs that are actually usable. A family whose bold is on disk
  /// and whose regular is not is *not* usable: the engine would letter every
  /// key in bold and the venue would have asked for neither.
  Future<Set<String>> _install(FontLibrary library, Set<String> wanted) async {
    if (library.fonts.isEmpty || wanted.isEmpty) return const <String>{};

    final Directory dir;
    try {
      dir = await _dir();
    } catch (_) {
      // No writable support directory. Nothing else here can work; plain
      // lettering, and no noise.
      return const <String>{};
    }

    final ready = <String>{};
    for (final font in library.fonts) {
      if (font.faces.isEmpty || !wanted.contains(font.slug)) continue;

      // Already handed to the engine in this run. Skipped before anything
      // touches the disk: a till hears a push every time a price changes, and
      // re-reading a megabyte of font files to conclude nothing had moved is
      // work done on a machine that is also taking money.
      if (_registered.contains(font.engineFamily)) {
        ready.add(font.slug);
        continue;
      }

      final bytes = <int, Uint8List>{};
      for (final face in font.faces) {
        final file = File(p.join(dir.path, face.fileName(font.slug)));
        try {
          if (file.existsSync() && file.lengthSync() > 0) {
            bytes[face.weight] = await file.readAsBytes();
            continue;
          }
          final res = await _client
              .get(Uri.parse(face.url))
              .timeout(const Duration(seconds: 20));
          if (res.statusCode != 200 || res.bodyBytes.isEmpty) continue;
          // Written to a temporary name and renamed, so a download cut off
          // halfway cannot leave a truncated file that every later start reads
          // as "already downloaded" and hands to the engine.
          final part = File('${file.path}.part');
          await part.writeAsBytes(res.bodyBytes, flush: true);
          await part.rename(file.path);
          bytes[face.weight] = res.bodyBytes;
        } catch (_) {
          // Offline, or this one face is unreachable. Carry on: the family is
          // only counted as ready if every face arrived.
        }
      }

      if (bytes.length != font.faces.length) continue;

      {
        try {
          _registered.add(font.engineFamily);
          final loader = FontLoader(font.engineFamily);
          // The engine picks a weight by reading each file's own metadata, so
          // the faces are added in whatever order and bold still resolves as
          // bold — but they are sorted anyway, because a family whose first
          // face is the bold one falls back to bold for a weight it has no
          // match for, and that fallback is where "everything is bold" comes
          // from.
          for (final weight in bytes.keys.toList()..sort()) {
            loader.addFont(
              Future.value(ByteData.view(bytes[weight]!.buffer)),
            );
          }
          await loader.load();
        } catch (_) {
          // A file that is not a font, or an engine that refused it. Take the
          // family back out so a later refresh can try again.
          _registered.remove(font.engineFamily);
          continue;
        }
      }
      ready.add(font.slug);
    }
    return ready;
  }

  /// Send a font file from this counter to the back office.
  ///
  /// On the terminal token, not on the office alone. Every other till route is
  /// a read; this one puts a file on the office's disk, and an unauthenticated
  /// write scoped by an email address is not a trade worth making.
  ///
  /// Throws with the server's own message on failure, because the server's
  /// message is the useful one — it is the only thing that can say "a .woff2
  /// works in a browser but not on a till".
  Future<void> upload({
    required String terminalToken,
    required File file,
    required String family,
    required int weight,
  }) async {
    final request = http.MultipartRequest(
      'POST',
      Uri.parse('$apiBase/api/till/fonts'),
    )
      ..headers['Authorization'] = 'Bearer $terminalToken'
      ..fields['family'] = family
      ..fields['weight'] = '$weight'
      ..files.add(await http.MultipartFile.fromPath('font', file.path));

    final streamed = await request.send().timeout(const Duration(seconds: 60));
    final res = await http.Response.fromStream(streamed);
    if (res.statusCode >= 200 && res.statusCode < 300) return;

    String message = 'That font would not upload.';
    try {
      final body = jsonDecode(res.body) as Map<String, dynamic>;
      if (body['error'] is String) message = body['error'] as String;
    } catch (_) {
      // A proxy's HTML error page, or nothing at all. The default reads better
      // than a stack trace at a counter.
    }
    throw Exception(message);
  }

  /// Change the font this venue's tills wear, from this till.
  Future<void> setVenueFont({
    required String terminalToken,
    required String? slug,
  }) async {
    final res = await _client
        .put(
          Uri.parse('$apiBase/api/till/font'),
          headers: {
            'Authorization': 'Bearer $terminalToken',
            'Content-Type': 'application/json',
          },
          body: jsonEncode({'fontFamily': slug ?? ''}),
        )
        .timeout(const Duration(seconds: 15));
    if (res.statusCode >= 200 && res.statusCode < 300) return;
    throw Exception('That font could not be set — ${res.statusCode}.');
  }
}

final fontsRepositoryProvider = Provider<FontsRepository>(
  (ref) => FontsRepository(apiBase: ref.watch(apiBaseProvider)),
);

/// The venue's fonts, downloaded and registered, refreshed when the back office
/// says so.
///
/// Listens for the same `screens` push a layout does, because that is what the
/// back office sends when a font is uploaded or removed — one message, one
/// handler, and no version of the till that ignores a second one.
///
/// **What gets downloaded is what is used, not what is on offer.** The venue's
/// own font plus every font named by a key on any programmed screen — which on
/// most venues is nothing at all, and on the rest is one or two families out of
/// eighteen. See [FontsRepository.load].
final fontsProvider = FutureProvider<FontLibrary>((ref) async {
  ref.listen(syncEventsProvider, (_, next) {
    final type = next.value?.type;
    if (type == 'screens' || type == 'till-settings') ref.invalidateSelf();
  });

  final office = ref.watch(officeProvider);
  if (office.isEmpty) return FontLibrary.empty;

  final wanted = <String>{};
  final venue = ref.watch(tillSettingsProvider).fontFamily;
  if (venue != null && venue.isNotEmpty) wanted.add(venue);
  // Every key that asks for a font of its own, across every screen — including
  // the ones a till has not opened yet. A page button leads to a screen the
  // clerk will reach mid-service, and fetching its lettering at the moment of
  // the tap is how a key changes shape in front of a customer.
  for (final screen in ref.watch(screensProvider).value?.screens ?? const []) {
    for (final button in screen.buttons) {
      final slug = button.fontFamily;
      if (slug != null && slug.isNotEmpty) wanted.add(slug);
    }
  }

  return ref.watch(fontsRepositoryProvider).load(office, wanted: wanted);
});

/// The family every key inherits, or null for the app's own typeface.
///
/// Null while the list is still loading, which is the honest answer: the
/// alternative is lettering the whole till in the app's font, then swapping the
/// lot a second later when the download finishes, in front of a customer.
final venueFontFamilyProvider = Provider<String?>((ref) {
  final library = ref.watch(fontsProvider).value;
  if (library == null) return null;
  final slug = ref.watch(tillSettingsProvider).fontFamily;
  return library.familyFor(slug);
});
